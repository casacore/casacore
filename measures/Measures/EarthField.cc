//# EarthField.cc:  EarthField class model calculations
//# Copyright (C) 1998-2000,2007
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes
#include <measures/Measures/EarthField.h>
#include <measures/Measures/MeasTable.h>
#include <casa/Quanta/MVPosition.h>
#include <casa/Arrays/Vector.h>
#include <casa/BasicSL/Constants.h>
#include <casa/BasicMath/Math.h>
#include <casa/System/AipsrcValue.h>

namespace casa { //# NAMESPACE CASA - BEGIN

//# Constants
const Double EarthField::INTV = 50000;

//# Static data
uInt EarthField::interval_reg_p = 0;

//# Constructors
EarthField::EarthField() :
  method_p(EarthField::STANDARD), fixedEpoch_p(MeasData::MJD2000), agh_p(0), 
  p_p(0), q_p(0), cl_p(0), sl_p(0),
  lres_p(0) {
    fillField();
}

EarthField::EarthField(const EarthField &other) {
  copy(other);
}

EarthField::EarthField(EarthFieldTypes model, Double catepoch) :
  method_p(model), fixedEpoch_p(catepoch),
  p_p(0), q_p(0), cl_p(0), sl_p(0),
  lres_p(0) {
    fillField();
  }

EarthField &EarthField::operator=(const EarthField &other) {
  if ( this != &other) copy(other);
  return *this;
}

void EarthField::init() {
  method_p = EarthField::STANDARD;
  fixedEpoch_p = MeasData::MJD2000;
  fillField();
}

void EarthField::init(EarthFieldTypes model, Double catepoch) {
  method_p = model;
  fixedEpoch_p = catepoch;
  fillField();
}

//# Destructor
EarthField::~EarthField() {}

//# Operators
// Calculate EarthField components
const Vector<Double> &EarthField::operator()(const MVPosition &pos) {
  calcField(pos);
  Vector<Double> dx((pos-checkPos_p).getValue());
  lres_p++; lres_p %= 4;
  for (Int i=0; i<3; i++) {
    result_p[lres_p](i) = pval_p[i] +
      dx(0)*dval_p[0][i] + dx(1)*dval_p[1][i] + dx(2)*dval_p[2][i];
  };
  return result_p[lres_p];
}

//# Member functions
const Vector<Double> *EarthField::derivative(const MVPosition &pos) {
  calcField(pos);
  lres_p=0;		// Make sure contiguous set
  for (Int j=0; j<3; j++) {
    lres_p++; lres_p %= 4;
    for (Int i=0; i<3; i++) {
      result_p[lres_p](i) = dval_p[j][i];
    };
  };
  return &result_p[1];
}

void EarthField::copy(const EarthField &other) {
  method_p = other.method_p;
  fixedEpoch_p = other.fixedEpoch_p;
  agh_p = other.agh_p;
  checkPos_p = other.checkPos_p;
  for (Int i=0; i<3; i++) {
    pval_p[i] = other.pval_p[i];
    for (Int k=0; k<3; k++) dval_p[i][k] = other.dval_p[i][k];
  };
  for (Int j=0; j<4; j++) {
    result_p[j] = other.result_p[j];
  };
}

void EarthField::fillField() {

  // Get the interpolation interval
  if (!EarthField::interval_reg_p) {
    interval_reg_p = 
      AipsrcValue<Double>::registerRC(String("measures.earthfield.d_interval"),
				      Unit("km"), Unit("m"),
				      EarthField::INTV);
  };

  checkPos_p = MVPosition(1e30, 1e30, 1e30);
  switch (method_p) {
  default:
    agh_p.resize(0);
    agh_p =  MeasTable::IGRF(fixedEpoch_p);
    p_p.resize(PQ_LEN);
    q_p.resize(PQ_LEN);
    cl_p.resize(2*PQ_LEN);
    sl_p.resize(2*PQ_LEN);
    break;
  };
  for (Int j=0; j<4; j++) {
    result_p[j].resize(3);
    for (Int k=0; k<3; ++k) result_p[j][k] = 0;
  };
  for (Int j=0; j<3; ++j) {
    pval_p[j] = 0;
    for (Int k=0; k<3; ++k) dval_p[j][k] = 0;
  };
}

void EarthField::refresh() {
  fillField();
}

void EarthField::calcField(const MVPosition &pos) {
  if (!pos.nearAbs(checkPos_p,
		   AipsrcValue<Double>::get(EarthField::interval_reg_p))) {
    checkPos_p = pos;
    Vector<Double> posmv(3);
    posmv = pos.getValue();
    Vector<Double> posv(3);
    posv = pos.get();
    switch (method_p) {
    case NONE: {
      for (uInt j=0; j<3; j++) {
	pval_p[j] =0;
	for (uInt i=0; i<3; i++) dval_p[j][i] =0;
      };
    }
    break;
    default: {
      Double slat, clat, slong, clong, x, y, z, ratio, rr(0), one, two, three;
      Int l, m, n, fn(0), fm, j, i;
      for (Int lp=0; lp<4; lp++) {
	slat = cos(C::pi_2 - posv(2));
	clat = sin(C::pi_2 - posv(2));
	slong = sin(posv(1));
	clong = cos(posv(1));
	cl_p(0) = clong;
	sl_p(0) = slong;
	x = 0.0;
	y = 0.0;
	z = 0.0;
	l = 0;
	m = 0;
	n = 0;
	ratio = 6371200/posv(0);
	//
	// Compute Schmidt quasi-normal coefficients P and X (=Q)
	//
	p_p(0) = 2.0 * slat;
	p_p(1) = 2.0 * clat;
	p_p(2) = 4.5 * slat * slat - 1.5;
	p_p(3) = 5.1961524 * clat * slat;
	q_p(0) = -clat;
	q_p(1) = slat;
	q_p(2) = -3.0 * clat * slat;
	q_p(3) = 1.7320508 * (slat * slat - clat * clat);
	
	for (Int k=0; k<PQ_LEN; k++) {
	  if (n-m-1 < 0) {
	    m = -1;
	    n++;
	    rr = pow(ratio, Double(n+2));
	    fn = n;
	  };
	  fm = m+1;
	  if (k-4 >=0) {
	    if (m+1-n == 0) {
	      one = sqrt(1.0 - 0.5/fm);
	      j = k - n - 1;
	      p_p(k) = (1.0 + 1.0/fm) * one * clat * p_p(j);
	      q_p(k) = one * (clat * q_p(j) + slat/fm * p_p(j));
	      sl_p(m) = sl_p(m-1) * cl_p(0) + cl_p(m-1) * sl_p(0);
	      cl_p(m) = cl_p(m-1) * cl_p(0)-sl_p(m-1) * sl_p(0);
	    } else {
	      one = sqrt(Double(fn * fn - fm * fm));
	      two = sqrt((fn-1.0) * (fn-1.0) - fm * fm)/one;
	      three = (2.0 * fn - 1.0)/one;
	      i = k-n;
	      j = k - 2 * n + 1;
	      p_p(k) = (fn+1.0) * (three * slat/fn * p_p(i) -
				   two/(fn-1.0) * p_p(j));
	      q_p(k) = three * (slat * q_p(i) - clat/fn * p_p(i)) -
		two * q_p(j);
	    };
	  };
	  //
	  // Synthesise X,Y,Z in geocentric coordinates
	  //
	  one = (agh_p(l)) * rr;
	  if (m == -1) {
	    x = x + one * q_p(k);
	    z = z - one * p_p(k);
	    l++;
	  } else {
	    two = (agh_p(l+1)) * rr;
	    three = one * cl_p(m) + two * sl_p(m);
	    x = x + three * q_p(k);
	    z = z - three * p_p(k);
	    if (clat > 0) {
	      y = y + (one * sl_p(m) -
		       two * cl_p(m)) * fm * p_p(k)/((fn+1.0) * clat);
	    } else {
	      y = y + (one * sl_p(m) -
		       two * cl_p(m)) * q_p(k) * slat;
	    };
	    l += 2;
	  };
	  m++;
	}; // calculation loop
	// Rotate from local vertical/meridian to ITRF one
	if (lp == 0) {
	  pval_p[0] = +x*slat*clong + z*clat*clong + y*slong;
	  pval_p[1] = -x*slat*slong + z*clat*slong - y*clong;
	  pval_p[2] = -x*clat + z*slat;
	} else {
	  dval_p[lp-1][0] = (+x*slat*clong + z*clat*clong + y*slong -
			     pval_p[0])/DER_INTV;
	  dval_p[lp-1][1] = (-x*slat*slong + z*clat*slong - y*clong -
			     pval_p[1])/DER_INTV;
	  dval_p[lp-1][2] = (-x*clat + z*slat -
			     pval_p[2])/DER_INTV;
	};
	if (lp < 3) {
	  if (lp != 0) posmv(lp-1) -= DER_INTV;
	  posmv(lp) += DER_INTV;
	  posv = MVPosition(posmv).get();
	};
      }; // derivative loop
    }
    break;
    };
  };
}

} //# NAMESPACE CASA - END

