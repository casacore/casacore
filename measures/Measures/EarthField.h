//# EarthField.h: EarthField class model claculations
//# Copyright (C) 1998
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//#
//# $Id$

#ifndef MEASURES_EARTHFIELD_H
#define MEASURES_EARTHFIELD_H

//# Includes
#include <casa/aips.h>
#include <casa/Arrays/Vector.h>
#include <casa/Quanta/MVPosition.h>

namespace casa { //# NAMESPACE CASA - BEGIN

//# Forward Declarations

//# Constants
// Length of P and Q arrays, half length of CL/SL arrays in IGRF model
const Int PQ_LEN = 65;
// Interval (m) for derivatives in IGRF model
const Double DER_INTV = 10000;

// <summary> EarthField class model calculations </summary>

// <use visibility=local>

// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="tEarthField" demos="">
// </reviewed>

// <prerequisite>
//   <li> <linkto class=Measure>Measure</linkto> class for use
//   <li> <linkto class=MeasTable>MeasTable</linkto> class for data
// </prerequisite>
//
// <etymology>
// Earth magnetic Field model
// </etymology>
//
// <synopsis>
// EarthField forms the class for Earth magnetic field calculations. It is a 
// simple container with the selected model, and the mean epoch.<br>
// The method is selected from one of the following:
// <ul>
//   <li> EarthField::STANDARD  (at 1998/05/18 the IGRF definition)
//   <li> EarthField::IGRF	(IGRF reference field model)
// </ul>
// Epochs can be specified as the MJD (with defined constants
// MeasData::MJD2000 and MeasData::MJD1950 or the actual MJD),
// leading to the following constructors:
// <ul>
//   <li> EarthField() default; assuming IGRF and MJD2000
//   <li> EarthField(method); assuming J2000 as epoch
//   <li> EarthField(method, epoch) with epoch Double(MJD)
// </ul>
// Actual EarthField for a certain position on Earth is calculated by the () 
// operator. Arguments can be:
// <ul>
//   <li> MVPosition: a position on Earth (in the ITRF frame)
// </ul>
// The returned value is a 3D vector of the field (in nT) in ITRF coordinates.
// The derivative (d<sup>-1</sup>) can be obtained as well by 
// derivative(MVPosition). <br>
// An EarthField can be re-initialised with a different method and/or other
// epoch with the <src>init()</src> functions (same format as constructors).
//
// To bypass the full, lengthy calculation actual returned values are calculated
// using the derivative if within about 50 km (error less than about
// 10<sup>-2</sup> G). A call to refresh() will re-initiate calculations
// from scratch.<br>
// The following details can be set with the 
// <linkto class=Aipsrc>Aipsrc</linkto> mechanism:
// <ul>
//  <li> measures.earthfield.d_interval: approximation radius
//	(km is default unit) over which a linear approximation
//	is used
// </ul>
// The field model is assumed to be constant over the time-span the class
// is used.
//
// The calculations are based on a routine provided by the IGRF community. See
// ftp.ngdc.noaa.gov/Solid_Earth/Mainfld_Mag/Models/IAGA, routine IGRFLIB.FOR.
// The values are in nT (10uG).
// </synopsis>
//
// <example>
//  <srcblock>
//	EarthField mine(EarthField::STANDARD,
//                      45837.0);		// define EarthField type
//						// for 84/05/17
//	MPosition pos;
//      MeasTable::Observatory(pos, "WSRT");	// Obervatory position
//  // Make sure correct position frame used
//      MVPosition x(MPosition::Convert(pos, MPosition::ITRF)().getValue());
//	MVEarthMagnetic now = mine(x);		// get EarthField
//  </srcblock>
// </example>
//
// <motivation>
// To have a container (with history) for field calculations
// </motivation>
//
// <todo asof="1998/05/18">
//   <li> nothing I know off
// </todo>

class EarthField {

public:

  //# Constants
  // Default interval to be used for linear approximation (in m)
  static const Double INTV;

  //# Enumerations
  // Known EarthField calculation models
  enum EarthFieldTypes {
    // Standard IGRF model
    IGRF,
    // Make the field equal to zero
    NONE,
    // Standard default model if none specified
    STANDARD = IGRF
  };

  //# Constructors
  // Default constructor, generates default J2000 EarthField identification
  EarthField();
  // Copy constructor
  EarthField(const EarthField &other);
  // Constructor with epoch in MJulian days (default is J2000) 
  explicit EarthField(EarthFieldTypes model, Double catepoch=51544.5);
  // Copy assignment
  EarthField &operator=(const EarthField &other);
  
  //# Destructor
  ~EarthField();
  
  //# Operators
  // Return the EarthField components. Note that the value returned has only
  // a lifetime as long as the EarthField container exists, and no new
  // derivative is asked for.
  const Vector<Double> &operator()(const MVPosition &pos);
  
  //# General Member Functions
  // Return derivatives of field (to X, Y, Z). Note that the value returned
  // has only a lifetime as long as the EarthField container exists, and
  // no new components or derivative is calculated. The returned value should
  // not be deleted.
  const Vector<Double> *derivative(const MVPosition &pos);
  // Re-initialise EarthField object with specified model and epoch, or
  // defaults STANDARD and J2000.
  // <group>
  void init();
  void init(EarthFieldTypes model, Double catepoch=51544.5);
  // </group>
  // Refresh calculations
  void refresh();
  
  private:
  //# Data members
  // Method to be used
  EarthFieldTypes method_p;
  // Fixed epoch to be used (MJD)
  Double fixedEpoch_p;
  // List of spherical components
  Vector<Double> agh_p;
  // Work arrays for calculations
  // <group>
  Vector<Double> p_p;
  Vector<Double> q_p;
  Vector<Double> cl_p;
  Vector<Double> sl_p;
  // </group>
  // Check position
  MVPosition checkPos_p;
  // Cached calculated field components
  Double pval_p[3];
  // Cached derivatives
  Double dval_p[3][3];
  // To reference results, and use a few in interim calculations, results are
  // calculated in a circular buffer.
  // Current result pointer
  Int lres_p;
  // Last calculation
  Vector<Double> result_p[4];
  // Interpolation interval
  static uInt interval_reg_p;

  //# Member functions
  // Make a copy
  void copy(const EarthField &other);
  // Create correct default fixedEpoch and catalogue field data
  void fillField();
  // Calculate EarthField for longitude and latitude and altitude (m)
  void calcField(const MVPosition &pos);

};
  

} //# NAMESPACE CASA - END

#endif


