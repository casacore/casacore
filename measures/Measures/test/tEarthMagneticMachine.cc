//# tEarthMagneticMachine.cc: This program tests the EarthMagneticMachine class
//# Copyright (C) 1998,1999,2000,2002
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This program is free software; you can redistribute it and/or modify it
//# under the terms of the GNU General Public License as published by the Free
//# Software Foundation; either version 2 of the License, or (at your option)
//# any later version.
//#
//# This program is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
//# more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with this program; if not, write to the Free Software Foundation, Inc.,
//# 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes
#include <casa/aips.h>
#include <casa/Exceptions/Error.h>
#include <measures/Measures.h>
#include <measures/Measures/EarthField.h>
#include <measures/Measures/EarthMagneticMachine.h>
#include <casa/Quanta/MVTime.h>
#include <measures/Measures/MDirection.h>
#include <measures/Measures/MPosition.h>
#include <measures/Measures/MEpoch.h>
#include <casa/Arrays/Vector.h>
#include <casa/Arrays/ArrayIO.h>
#include <casa/iostream.h>

#include <casa/namespace.h>
int main() {

  try {
    cout << "Test Earth Magnetic field machine" << endl;
    cout << "--------------------------------------" << endl;
    MVTime dat(1998,5,18);
    MVPosition mvobs(Quantity(3828488.86, "m").getBaseValue(),
		     Quantity(443253.42, "m").getBaseValue(),
		     Quantity(5064977.78, "m").getBaseValue());
    MPosition obs(mvobs);
    MeasFrame frame((MEpoch(MVEpoch(dat.day()))), obs);
    
    cout << "Date:      " << dat.string(MVTime::YMD +
					MVTime::NO_TIME, 6) <<
      endl;
    cout << "Position:  " << obs.getValue().get() << endl;
    cout << "           " << obs.getAngle("deg") << endl;
    cout << "           " << obs.getValue().getLength("km") << endl;
    
    EarthField ef(EarthField::STANDARD, dat.day());
    cout << "Result:    " << ef(obs.getValue()) << endl;
    
    cout << "----------- H=0 -- El=90 ------" << endl;
    {
      MDirection::Ref mvref(MDirection::ITRF, frame);
      MVDirection mvd(obs.getValue());
      EarthMagneticMachine fm(mvref, Quantum<Double>(0, "km"),
			      frame);
      fm.calculate(mvd);
      cout << "LOS:           " << fm.getLOSField() << endl;
      cout << "LOS:           " << fm.getLOSField("G") << endl;
      cout << "Long:          " << fm.getLong() << endl;
      cout << "Long:          " << fm.getLong("deg") << endl;
      cout << "LOS:           " << fm() << endl;
      cout << "LOS:           " << fm("G") << endl;
    }
    cout << "----------- H=0 -- El=90 ------" << endl;
    {
      MDirection::Ref mvref(MDirection::AZEL, frame);
      MVDirection mvd(Quantity(0, "deg"), Quantity(90, "deg"));
      EarthMagneticMachine fm(mvref, Quantum<Double>(0, "km"),
			      obs, MEpoch(MVEpoch(dat.day())));
      fm.calculate(mvd);
      cout << "LOS:           " << fm.getLOSField() << endl;
      cout << "LOS:           " << fm.getLOSField("G") << endl;
      cout << "Long:          " << fm.getLong() << endl;
      cout << "Long:          " << fm.getLong("deg") << endl;
      cout << "LOS:           " << fm() << endl;
      cout << "LOS:           " << fm("G") << endl;
    }
     cout << "----------- Along field ------" << endl;
    {
      MDirection::Ref mvref(MDirection::AZEL, frame);
      Vector<Double> xvd(3);
      xvd(0) = 18312;
      xvd(1) =  -381;
      xvd(2) = 45184;
      MVDirection mvd(xvd);
      EarthMagneticMachine fm(mvref, Quantum<Double>(0, "km"),
			      obs, MEpoch(MVEpoch(dat.day())));
      fm.calculate(mvd);
      cout << "LOS:           " << fm.getLOSField() << endl;
      cout << "LOS:           " << fm.getLOSField("G") << endl;
      cout << "Long:          " << fm.getLong() << endl;
      cout << "Long:          " << fm.getLong("deg") << endl;
      cout << "LOS:           " << fm() << endl;
      cout << "LOS:           " << fm("G") << endl;
    }
   cout << "----------- Ha=0 -- Dec= 52.7316 ------" << endl;
    {
      MDirection::Ref mvref(MDirection::HADEC, frame);
      MVDirection mvd(Quantity(0, "deg"), Quantity( 52.7316, "deg"));
      EarthMagneticMachine fm(mvref, Quantum<Double>(0, "km"),
			      obs, MEpoch(MVEpoch(dat.day())));
      fm.calculate(mvd);
      cout << "LOS:           " << fm.getLOSField() << endl;
      cout << "LOS:           " << fm.getLOSField("G") << endl;
      cout << "Long:          " << fm.getLong() << endl;
      cout << "Long:          " << fm.getLong("deg") << endl;
      cout << "LOS:           " << fm() << endl;
      cout << "LOS:           " << fm("G") << endl;
    }
    cout << "----------- H=0 -- El=45 ------" << endl;
    {
      MDirection::Ref mvref(MDirection::AZEL, frame);
      MVDirection mvd(Quantity(0, "deg"), Quantity(45, "deg"));
      EarthMagneticMachine fm(mvref, Quantum<Double>(0, "km"),
			      frame);
      fm.calculate(mvd);
      cout << "LOS:           " << fm.getLOSField() << endl;
      cout << "LOS:           " << fm.getLOSField("G") << endl;
      cout << "Long:          " << fm.getLong() << endl;
      cout << "Long:          " << fm.getLong("deg") << endl;
      cout << "LOS:           " << fm() << endl;
      cout << "LOS:           " << fm("G") << endl;
    }
    cout << "----------- H=200 -- El=45 ------" << endl;
    {
      MDirection::Ref mvref(MDirection::AZEL, frame);
      MVDirection mvd(Quantity(0, "deg"), Quantity(45, "deg"));
      EarthMagneticMachine fm(mvref, Quantum<Double>(200, "km"),
			      frame);
      fm.calculate(mvd);
      cout << "LOS:           " << fm.getLOSField() << endl;
      cout << "LOS:           " << fm.getLOSField("G") << endl;
      cout << "Long:          " << fm.getLong() << endl;
      cout << "Long:          " << fm.getLong("deg") << endl;
      cout << "LOS:           " << fm() << endl;
      cout << "LOS:           " << fm("G") << endl;
    }
    cout << "----------- H=200 -- El=0 -- N ----" << endl;
    {
      MDirection::Ref mvref(MDirection::AZEL, frame);
      MVDirection mvd(Quantity(0, "deg"), Quantity(0, "deg"));
      EarthMagneticMachine fm(mvref, Quantum<Double>(200, "km"),
			      frame);
      fm.calculate(mvd);
      cout << "LOS:           " << fm.getLOSField() << endl;
      cout << "LOS:           " << fm.getLOSField("G") << endl;
      cout << "Long:          " << fm.getLong() << endl;
      cout << "Long:          " << fm.getLong("deg") << endl;
      cout << "LOS:           " << fm() << endl;
      cout << "LOS:           " << fm("G") << endl;
    }
    cout << "----------- H=200 -- El=5 -- E ----" << endl;
    {
      MDirection::Ref mvref(MDirection::AZEL, frame);
      MVDirection mvd(Quantity(90, "deg"), Quantity(5, "deg"));
      EarthMagneticMachine fm(mvref, Quantum<Double>(200, "km"),
			      frame);
      fm.calculate(mvd);
      cout << "LOS:           " << fm.getLOSField() << endl;
      cout << "LOS:           " << fm.getLOSField("G") << endl;
      cout << "Long:          " << fm.getLong() << endl;
      cout << "Long:          " << fm.getLong("deg") << endl;
      cout << "LOS:           " << fm() << endl;
      cout << "LOS:           " << fm("G") << endl;
    }
    cout << "----------- H=200 -- El=5 -- W ----" << endl;
    {
      MDirection::Ref mvref(MDirection::AZEL, frame);
      MVDirection mvd(Quantity(-90, "deg"), Quantity(5, "deg"));
      EarthMagneticMachine fm(mvref, Quantum<Double>(200, "km"),
			      frame);
      fm.calculate(mvd);
      cout << "LOS:           " << fm.getLOSField() << endl;
      cout << "LOS:           " << fm.getLOSField("G") << endl;
      cout << "Long:          " << fm.getLong() << endl;
      cout << "Long:          " << fm.getLong("deg") << endl;
      cout << "LOS:           " << fm() << endl;
      cout << "LOS:           " << fm("G") << endl;
      cout << "------------- recalculate ----------" << endl;
      fm.reCalculate();
      cout << "LOS:           " << fm(mvd, "G") << endl;
      cout << "------------- use set() ------------" << endl;
      EarthMagneticMachine fm0;
      fm0.set(mvref);
      fm0.set(Quantity(200, "km"));
      fm0.set(frame);
      fm0.calculate(mvd);
      cout << "LOS:           " << fm0("G") << endl;
      EarthMagneticMachine fm1;
      fm1 = fm0;
      fm1.calculate(mvd);
      cout << "LOS:           " << fm1("G") << endl;
      cout << "LOS:           " << fm1(mvd) << endl;
      cout << "LOS:           " << fm1(mvd, "G") << endl;
      cout << "Long:          " << fm1.getLong(mvd) << endl;
      cout << "Long:          " << fm1.getLong(mvd, "deg") << endl;
      cout << "Field:         " << fm1.getField() << endl;
      cout << "Field:         " << fm1.getField(mvd) << endl;
      cout << "Pos:           " << fm1.getPosition() << endl;
      cout << "Pos:           " << fm1.getPosition(mvd) << endl;
      EarthMagneticMachine fm2(fm1);
      fm2.calculate(mvd);
      cout << "LOS:           " << fm2("G") << endl;
      cout << "------------- iterate height ------------" << endl;
      Quantum<Double> qhgt(200, "km"); 
      cout << "LOS:           " << fm1("G") << endl;
      cout << "LOS:           " << fm1(qhgt) << endl;
      cout << "LOS:           " << fm1(qhgt, "G") << endl;
      cout << "Long:          " << fm1.getLong() << endl;
      cout << "Long:          " << fm1.getLong("deg") << endl;
      cout << "Field:         " << fm1.getField() << endl;
      cout << "Pos:           " << fm1.getPosition() << endl;
      qhgt = Quantum<Double>(190, "km");
      cout << "LOS 190:       " << fm1(qhgt, "G") << endl;
      qhgt = Quantum<Double>(200, "km");
      cout << "LOS 200:       " << fm1(qhgt, "G") << endl;
      qhgt = Quantum<Double>(210, "km");
      cout << "LOS 210:       " << fm1(qhgt, "G") << endl;
    }
    
  } catch (AipsError x) {
    cout << x.getMesg() << endl;
  } 
  
  return 0;
}
