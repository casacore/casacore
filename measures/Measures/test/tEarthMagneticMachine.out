Test Earth Magnetic field machine
--------------------------------------
Date:      1998/05/18
Position:  [6.36457e+06, 0.115264, 0.92034]
           [6.60417, 52.7316] deg
           6364.57 km
Result:    [41647.4, 1813.7, 24906]
----------- H=0 -- El=90 ------
LOS:           44998.9
LOS:           0.449989 G
Long:          0.115264
Long:          6.60417 deg
LOS:           44998.9
LOS:           0.449989 G
----------- H=0 -- El=90 ------
LOS:           44998.9
LOS:           0.449989 G
Long:          0.115264
Long:          6.60417 deg
LOS:           44998.9
LOS:           0.449989 G
----------- Along field ------
LOS:           34962.8
LOS:           0.349628 G
Long:          0.115264
Long:          6.60417 deg
LOS:           34962.8
LOS:           0.349628 G
----------- Ha=0 -- Dec= 52.7316 ------
LOS:           44999
LOS:           0.44999 G
Long:          0.115264
Long:          6.60417 deg
LOS:           44999
LOS:           0.44999 G
----------- H=0 -- El=45 ------
LOS:           19085.8
LOS:           0.190858 G
Long:          0.115264
Long:          6.60417 deg
LOS:           19085.8
LOS:           0.190858 G
----------- H=200 -- El=45 ------
LOS:           19600.9
LOS:           0.196009 G
Long:          0.115264
Long:          6.60417 deg
LOS:           19600.9
LOS:           0.196009 G
----------- H=200 -- El=0 -- N ----
LOS:           864.604
LOS:           0.00864604 G
Long:          0.115264
Long:          6.60417 deg
LOS:           864.604
LOS:           0.00864604 G
----------- H=200 -- El=5 -- E ----
LOS:           3048.74
LOS:           0.0304874 G
Long:          0.39911
Long:          22.8673 deg
LOS:           3048.74
LOS:           0.0304874 G
----------- H=200 -- El=5 -- W ----
LOS:           7557.64
LOS:           0.0755764 G
Long:          -0.168581
Long:          -9.659 deg
LOS:           7557.64
LOS:           0.0755764 G
------------- recalculate ----------
LOS:           0.0755764 G
------------- use set() ------------
LOS:           0.0755764 G
LOS:           0.0755764 G
LOS:           7557.64
LOS:           0.0755764 G
Long:          -0.168581
Long:          -9.659 deg
Field:         [38718.3, 377.355, 21091.1]
Field:         [38718.3, 377.355, 21091.1]
Pos:           [4.01992e+06, -684178, 5.14449e+06]
Pos:           [4.01992e+06, -684178, 5.14449e+06]
LOS:           0.0755764 G
------------- iterate height ------------
LOS:           0.0755764 G
LOS:           7557.64
LOS:           0.0755764 G
Long:          -0.168581
Long:          -9.659 deg
Field:         [38718.3, 377.355, 21091.1]
Pos:           [4.01992e+06, -684178, 5.14449e+06]
LOS 190:       0.0755459 G
LOS 200:       0.0755764 G
LOS 210:       0.0756062 G
